File: OJ practices/LeetCode题目集/690. Employee Importance(easy).java
import java.util.HashMap;
import java.util.Map;

/*
 * @Autor: SourDumplings
 * @Date: 2019-10-17 22:04:27
 * @Link: https://github.com/SourDumplings/
 * @Email: changzheng300@foxmail.com
 * @Description: https://leetcode.com/problems/employee-importance/
 */

/*
// Employee info
class Employee {
   // It's the unique id of each node;
   // unique id of this employee
   public int id;
   // the importance value of this employee
   public int importance;
   // the id of direct subordinates
   public List<Integer> subordinates;
};
*/
class Solution
{
    private int res;

    public int getImportance(List<Employee> employees, int id)
    {
        res = 0;
        Map<Integer, Employee> map = new HashMap<>();
        for (Employee e : employees)
        {
            map.put(e.id, e);
        }
        calcImportance(map, id);
        return res;
    }

    private void calcImportance(Map<Integer, Employee> map, int id)
    {
        Employee e = map.get(id);
        res += e.importance;
        for (int i : e.subordinates)
        {
            calcImportance(map, i);
        }
    }
}

File: OJ practices/LeetCode题目集/690. Employee Importance(easy).cpp
/*
 * @Author: SourDumplings
 * @Date: 2019-10-17 21:52:07
 * @Link: https://github.com/SourDumplings/
 * @Email: changzheng300@foxmail.com
 * @Description: https://leetcode.com/problems/employee-importance/
 */

/*
// Employee info
class Employee {
public:
    // It's the unique ID of each node.
    // unique id of this employee
    int id;
    // the importance value of this employee
    int importance;
    // the id of direct subordinates
    vector<int> subordinates;
};
*/
class Solution
{
public:
    int getImportance(vector<Employee *> employees, int id)
    {
        int res = 0;
        unordered_map<int, Employee *> m;
        for (auto &&e : employees)
        {
            m[e->id] = e;
        }
        calcImportance(m, id, res);
        return res;
    }

    void calcImportance(unordered_map<int, Employee *> &m, int id, int &res)
    {
        Employee *e = m[id];
        res += e->importance;
        for (auto &&i : e->subordinates)
        {
            calcImportance(m, i, res);
        }
    }
};